### PR TITLE
processor (ticdc): fix a panic and data race in processor initializing.

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/7805

#### Tests 

 - Integration test
 - Manual test (add detailed scripts or steps below)

### Release note 

```release-note
Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

File: cdc/processor/sourcemanager/engine/factory/factory.go [hunk 1]
```diff
@@ -68,11 +71,12 @@ func (f *SortEngineFactory) Create(ID model.ChangeFeedID) (e engine.SortEngine,
 		if e, exists = f.engines[ID]; exists {
 			return e, nil
 		}
-		if len(f.dbs) == 0 {
+		if !f.dbInitialized.Load() && len(f.dbs) == 0 {
```
Comment (hicqu) [inline cdc/processor/sourcemanager/engine/factory/factory.go:74]: db.Initialized must be false here.

File: cdc/processor/sourcemanager/engine/factory/factory.go [hunk 2]
```diff
@@ -54,6 +55,8 @@ type SortEngineFactory struct {
 	pebbleConfig *config.DBConfig
 	dbs          []*pebble.DB
 	writeStalls  []writeStall
+
+	dbInitialized *atomic.Bool
```
Comment (hicqu) [inline cdc/processor/sourcemanager/engine/factory/factory.go:59]: Please comment why it's necessary.

**Comment by hicqu** [comment]
Rest LGTM.

File: cdc/processor/processor.go [hunk 3]
```diff
@@ -650,7 +650,12 @@ func (p *processor) Tick(ctx cdcContext.Context) error {
 	}
 
 	p.metricProcessorTickDuration.Observe(costTime.Seconds())
-	p.refreshMetrics()
+
+	// we should check if this error is nil,
+	// otherwise the function called below may panic.
+	if err != nil {
```
Comment (3AceShowHand) [inline cdc/processor/processor.go:656]: I think we can check error immediately after the `tick`.

File: cdc/processor/sourcemanager/engine/factory/factory.go (hunk 2 above)
Comment (3AceShowHand) [reply cdc/processor/sourcemanager/engine/factory/factory.go:59]: just name it `initialized` should be ok.

**Comment by asddongmen** [comment]
https://github.com/pingcap/tiflow/pull/7808 contains this PR.

**Pull request closed without merging**